README.md
Turns any path into a dashed path.

# Preview

<img src="https://user-images.githubusercontent.com/57846029/224071734-bddbfaa0-49af-4b18-bfcf-fa40d46a8752.png"  width="400">

# Installation

Add this to your package's `pubspec.yaml` file:

```
dependencies:
  dashed_path: ^1.0.0
```

# Import

```dart
import 'package:dashed_path/dashed_path.dart';
```

# Usage

Basic usage example corresponding to the above shown preview:

```dart
CustomPaint(
painter: DashedPathPainter(
    originalPath: Path()..lineTo(100, 0),
    pathColor: Colors.red,
    strokeWidth: 5.0,
    dashGapLength: 10.0,
    dashLength: 10.0,
),
size: const Size(100.0, 2.0),
),
const SizedBox(height: 50),
CustomPaint(
painter: DashedPathPainter(
    originalPath: Path()
    ..addOval(
        const Rect.fromLTWH(0, 0, 100, 100),
    ),
    pathColor: Colors.white,
),
size: const Size(100.0, 100.0),
),
const SizedBox(height: 50),
CustomPaint(
painter: DashedPathPainter(
    originalPath: Path()
    ..addRect(
        const Rect.fromLTWH(0, 0, 100, 100),
    ),
    pathColor: Colors.grey,
    strokeWidth: 2.0,
    dashLength: 25.0,
),
size: const Size(100.0, 100.0),
)
```
Have a look at the example folder for a complete usage example.

You can define the following properties:
| Property | Description | 
| -------- | ----------- |
| originalPath | The original path that should be converted to a dashed path. |
| pathColor | The path's color. |
| strokeWidth | The stroke width of the path. |
| dashGapLength | The gap length between the dashes.  |
| dashLength | The length of each dash. |


CHANGELOG.md
## 1.0.0

* Initial Release

## 1.0.1

* Updated README


test/dashed_path_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:dashed_path/dashed_path.dart';

void main() {
  testWidgets('Test correct dash creation', (WidgetTester tester) async {
    await tester.pumpWidget(const _DashedPathExampleApp());
    await expectLater(find.byType(_DashedPathExampleApp),
        matchesGoldenFile('example_dashes.png'));
  });
}

class _DashedPathExampleApp extends StatelessWidget {
  const _DashedPathExampleApp();

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Dashed Path Example App',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        scaffoldBackgroundColor: const Color.fromARGB(255, 33, 34, 35),
      ),
      home: const _ExampleDashedPathPainter(),
    );
  }
}

class _ExampleDashedPathPainter extends StatelessWidget {
  const _ExampleDashedPathPainter({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: SingleChildScrollView(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              const SizedBox(height: 25),
              CustomPaint(
                painter: DashedPathPainter(
                  originalPath: Path()..lineTo(100, 0),
                  pathColor: Colors.red,
                  strokeWidth: 5.0,
                  dashGapLength: 10.0,
                  dashLength: 10.0,
                ),
                size: const Size(100.0, 2.0),
              ),
              const SizedBox(height: 50),
              CustomPaint(
                painter: DashedPathPainter(
                  originalPath: Path()
                    ..addOval(
                      const Rect.fromLTWH(0, 0, 100, 100),
                    ),
                  pathColor: Colors.white,
                ),
                size: const Size(100.0, 100.0),
              ),
              const SizedBox(height: 50),
              CustomPaint(
                painter: DashedPathPainter(
                  originalPath: Path()
                    ..addRect(
                      const Rect.fromLTWH(0, 0, 100, 100),
                    ),
                  pathColor: Colors.grey,
                  strokeWidth: 2.0,
                  dashLength: 25.0,
                ),
                size: const Size(100.0, 100.0),
              ),
              const SizedBox(height: 25),
            ],
          ),
        ),
      ),
    );
  }
}


lib/dashed_path.dart
import 'dart:ui' as ui;
import 'dart:math' as math;

import 'package:flutter/material.dart';

/// A painter that turns any path into a dashed path.
class DashedPathPainter extends CustomPainter {
  /// The path that you should be dashed.
  ///
  /// It can be of any shape (e.g. circle, straight line, rectangle).
  final Path originalPath;

  /// The color that will be applied to the dashed path.
  final Color pathColor;

  /// The path's width.
  final double strokeWidth;

  /// The length of the dash.
  final double dashLength;

  /// The length of the gap between two dashes.
  final double dashGapLength;

  late DashedPathProperties _dashedPathProperties;

  /// A DashedPathPainter that converts any path into a dashed path.
  ///
  /// That means you can convert any path shape you like (e.g. circle, rectangle, lines, custom paths) into a dashed version.
  DashedPathPainter({
    required this.originalPath,
    required this.pathColor,
    this.strokeWidth = 3.0,
    this.dashGapLength = 5.0,
    this.dashLength = 10.0,
  })  : assert(dashLength > 0.0, 'dashLength must be > 0.0'),
        assert(dashGapLength > 0.0, 'dashGapLength must be > 0.0');

  @override
  void paint(Canvas canvas, Size size) {
    _dashedPathProperties = DashedPathProperties(
      path: Path(),
      dashLength: dashLength,
      dashGapLength: dashGapLength,
    );
    final dashedPath = _getDashedPath(originalPath, dashLength, dashGapLength);
    canvas.drawPath(
      dashedPath,
      Paint()
        ..style = PaintingStyle.stroke
        ..color = pathColor
        ..strokeWidth = strokeWidth,
    );
  }

  @override
  bool shouldRepaint(DashedPathPainter oldDelegate) =>
      oldDelegate.originalPath != originalPath ||
      oldDelegate.pathColor != pathColor ||
      oldDelegate.strokeWidth != strokeWidth ||
      oldDelegate.dashGapLength != dashGapLength ||
      oldDelegate.dashLength != dashLength;

  Path _getDashedPath(
      Path originalPath, double dashLength, double dashGapLength) {
    final metricsIterator = originalPath.computeMetrics().iterator;
    while (metricsIterator.moveNext()) {
      final metric = metricsIterator.current;
      _dashedPathProperties.extractedPathLength = 0.0;
      while (_dashedPathProperties.extractedPathLength < metric.length) {
        if (_dashedPathProperties.addDashNext) {
          _dashedPathProperties.addDash(metric, dashLength);
        } else {
          _dashedPathProperties.addDashGap(metric, dashGapLength);
        }
      }
    }
    return _dashedPathProperties.path;
  }
}

class DashedPathProperties {
  double extractedPathLength;
  Path path;

  final double _dashLength;
  double _remainingDashLength;
  double _remainingDashGapLength;
  bool _previousWasDash = false;

  DashedPathProperties({
    required this.path,
    required double dashLength,
    required double dashGapLength,
  })  : assert(dashLength > 0.0, 'dashLength must be > 0.0'),
        assert(dashGapLength > 0.0, 'dashGapLength must be > 0.0'),
        _dashLength = dashLength,
        _remainingDashLength = dashLength,
        _remainingDashGapLength = dashGapLength,
        _previousWasDash = false,
        extractedPathLength = 0.0;

  bool get addDashNext {
    if (_remainingDashLength != _dashLength || !_previousWasDash) {
      return true;
    }
    return false;
  }

  void addDash(ui.PathMetric metric, double dashLength) {
    // Calculate lengths (actual + available)
    final end = _calculateLength(metric, _remainingDashLength);
    final availableEnd = _calculateLength(metric, dashLength);
    // Add path
    final pathSegment = metric.extractPath(extractedPathLength, end);
    path.addPath(pathSegment, Offset.zero);
    // Update
    final delta = _remainingDashLength - (end - extractedPathLength);
    _remainingDashLength = _updateRemainingLength(
      delta: delta,
      end: end,
      availableEnd: availableEnd,
      initialLength: dashLength,
    );
    extractedPathLength = end;
    _previousWasDash = true;
  }

  void addDashGap(ui.PathMetric metric, double dashGapLength) {
    // Calculate lengths (actual + available)
    final end = _calculateLength(metric, _remainingDashGapLength);
    final availableEnd = _calculateLength(metric, dashGapLength);
    // Move path's end point
    ui.Tangent tangent = metric.getTangentForOffset(end)!;
    path.moveTo(tangent.position.dx, tangent.position.dy);
    // Update
    final delta = end - extractedPathLength;
    _remainingDashGapLength = _updateRemainingLength(
      delta: delta,
      end: end,
      availableEnd: availableEnd,
      initialLength: dashGapLength,
    );
    extractedPathLength = end;
    _previousWasDash = false;
  }

  double _calculateLength(ui.PathMetric metric, double addedLength) {
    return math.min(extractedPathLength + addedLength, metric.length);
  }

  double _updateRemainingLength({
    required double delta,
    required double end,
    required double availableEnd,
    required double initialLength,
  }) {
    return (delta > 0 && availableEnd == end) ? delta : initialLength;
  }
}


example/lib/main.dart
import 'package:dashed_path/dashed_path.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(const DashedPathExampleApp());
}

class DashedPathExampleApp extends StatelessWidget {
  const DashedPathExampleApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Dashed Path Example App',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        scaffoldBackgroundColor: const Color.fromARGB(255, 33, 34, 35),
      ),
      home: const ExampleDashedPathPainter(),
    );
  }
}

class ExampleDashedPathPainter extends StatelessWidget {
  const ExampleDashedPathPainter({Key? key}) : super(key: key);

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Center(
        child: SingleChildScrollView(
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.center,
            children: [
              const SizedBox(height: 25),
              CustomPaint(
                painter: DashedPathPainter(
                  originalPath: Path()..lineTo(100, 0),
                  pathColor: Colors.red,
                  strokeWidth: 5.0,
                  dashGapLength: 10.0,
                  dashLength: 10.0,
                ),
                size: const Size(100.0, 2.0),
              ),
              const SizedBox(height: 50),
              CustomPaint(
                painter: DashedPathPainter(
                  originalPath: Path()
                    ..addOval(
                      const Rect.fromLTWH(0, 0, 100, 100),
                    ),
                  pathColor: Colors.white,
                ),
                size: const Size(100.0, 100.0),
              ),
              const SizedBox(height: 50),
              CustomPaint(
                painter: DashedPathPainter(
                  originalPath: Path()
                    ..addRect(
                      const Rect.fromLTWH(0, 0, 100, 100),
                    ),
                  pathColor: Colors.grey,
                  strokeWidth: 2.0,
                  dashLength: 25.0,
                ),
                size: const Size(100.0, 100.0),
              ),
              const SizedBox(height: 25),
            ],
          ),
        ),
      ),
    );
  }
}
